chore: add readme for auth with examples (#703)
aws-amplify/amplify-backend: packages/auth-construct/README.md

# AmplifyAuth Construct

> This package vends an L3 CDK Construct wrapping the behavior of Amplify Auth. This enables quick development and interation with AWS Cognito.

Reviewer: ```suggestion
This package vends an L3 CDK Construct that enables faster, easier and secure app authentication and authorization, powered by Amazon Cognito. Amplify Auth lets you quickly set up secure authentication flows with a fully-managed user directory. Control what users have access to in your mobile and web apps with Amplify Auth's built-in authorization capabilities.
```

> This package vends an L3 CDK Construct that enables faster, easier and secure app authentication and authorization powered by Amazon Cognito. Amplify Auth lets you quickly set up secure authentication flows with a fully-managed user directory. Control what users have access to in your mobile and web apps with Amplify Auth's built-in authorization capabilities.

Reviewer 2: nit: I always forget what "L3 CDK Construct" is. Would be helpful if we could add a link https://docs.aws.amazon.com/cdk/v2/guide/constructs.html

Author: im not sure about this one since data team did not include links to other cdk docs https://constructs.dev/packages/@aws-amplify/data-construct/v/1.4.2?lang=typescript

The primary way to use this construct is to invoke it with a configuration object. You can declare the individual settings for your authentication by passing them as properties to the AmplifyAuth construct.

Note: only a single instance of the `AmplifyAuth` construct can be invoked within a CDK synthesis at this point in time.

## Examples

### Simple email login with default settings

> In this example, you create a simple stack with email login enabled (by default). Deploying this will create a UserPool, UserPoolClient, IdentityPool, and Authenticated/Unauthenticated Roles.

Reviewer: ```suggestion
In this example, you will create a simple stack with email login enabled (by default). Deploying this will create a UserPool, UserPoolClient, IdentityPool, and Authenticated/Unauthenticated IAM Roles.
```

Author: fixed

```ts
import { App, Stack } from 'aws-cdk-lib';
import { AmplifyAuth } from '@aws-amplify/auth-construct';

const app = new App();

> const stack = new Stack(app, 'Auth');

Reviewer: ```suggestion
const stack = new Stack(app, 'AuthStack');
```

Author: updated

> In this example, you will create a simple stack with email login enabled (by default). Deploying this will create a UserPool, UserPoolClient, IdentityPool, and Authenticated/Unauthenticated IAM Roles.

Reviewer 2: Should that be `a UserPool, a UserPoolClient, an IdentityPool, and an Authenticated/Unauthenticated IAM Role." or "a UserPool, a UserPoolClient, an IdentityPool, and Authenticated/Unauthenticated IAM Roles."?

Author: looks valid from a few grammar check sites

new AmplifyAuth(stack, 'Auth');
```

### Email login with customized multifactor settings

> In this example, we create a simple stack with email login enabled and with customized multifactor authentication settings.

Reviewer: ```suggestion
In this example, you will create a simple stack with email login enabled and with customized multi factor authentication (MFA) settings.
```

Author: fixed

```ts
import { App, Stack } from 'aws-cdk-lib';
import { AmplifyAuth } from '@aws-amplify/auth-construct';

const app = new App();

> const stack = new Stack(app, 'Auth');

Reviewer: ```suggestion
const stack = new Stack(app, 'AuthStack');
```

Author: fixed

new AmplifyAuth(stack, 'Auth', {
  loginWith: {
    email: true,
  },
  multifactor: {
    mode: 'OPTIONAL',
    sms: {

>       smsMessage: (code: string) => `valid MFA message with ${code}`,

Reviewer: Is this the right sample message here?

Author: updated it to `Your verification code is ${code}.`

    },
    totp: false,
  },
});
```

### Customized email and phone login with external login providers

> In this example, we create a stack with email, phone, and external login providers. Additionally, we have customized email and phone verification messages.

Reviewer: ```suggestion
In this example, you will create a stack with email, phone, and external login providers. Additionally, you can customize email and phone verification messages.
```

Author: fixed

```ts
import { App, Stack, SecretValue } from 'aws-cdk-lib';
import { AmplifyAuth } from '@aws-amplify/auth-construct';
import { UserPoolIdentityProviderSamlMetadataType } from 'aws-cdk-lib/aws-cognito';

const app = new App();

> const stack = new Stack(app, 'Auth');

Reviewer: ```suggestion
const stack = new Stack(app, 'AuthStack');
```

Author: fixed

new AmplifyAuth(stack, 'Auth', {
  loginWith: {
    email: {
      verificationEmailStyle: 'CODE',
      verificationEmailBody: (code: string) =>
        `Your verification code is ${code}.`,
      verificationEmailSubject: 'My custom email subject',
    },
    phone: {
      verificationMessage: (code: string) =>
        `Your verification code is ${code}.`,
    },
    externalProviders: {
      google: {
        clientId: 'googleClientId',

>         clientSecret: SecretValue.unsafePlainText(googleClientSecret),

Reviewer: Add a cdk link on how to configure/use these secrets

Author: there is really good hover text on SecretValue with a link
![image](https://github.com/aws-amplify/amplify-backend/assets/110861985/b27e5534-5fdf-4ca3-93d5-5a6a63be7e5f)

Reviewer: This documentation link should be added here. Not often folks are reading Readme in an IDE.

Author: good point; i added a link to https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.SecretValue.html

      },
      facebook: {
        clientId: 'facebookClientId',
        clientSecret: 'facebookClientSecret',
      },
      signInWithApple: {
        clientId: 'appleClientId',
        keyId: 'appleKeyId',
        privateKey: 'applePrivateKey',
        teamId: 'appleTeamId',
      },
      loginWithAmazon: {
        clientId: 'amazonClientId',

>         clientSecret: 'amazonClientSecret',

Reviewer: Shouldn't all secrets be shown coming from `SecretValue`

Author: I'll double check, this is for the CDK construct readme

Author: SecretValue is not provided to cdk users, so won't include in these examples

      },
      oidc: {
        clientId: 'oidcClientId',
        clientSecret: 'oidcClientSecret',
        issuerUrl: 'oidcIssuerUrl',
        name: 'oidcProviderName',
      },
      saml: {
        name: 'samlProviderName',
        metadata: {
          metadataContent: 'samlMetadataContent',
          metadataType: UserPoolIdentityProviderSamlMetadataType.FILE,
        },
      },
    },
  },
});
```

### Custom user attributes

> In this example, you are customizing the set of attributes that are required for every user in the UserPool.

Reviewer: ```suggestion
In this example, you will customize the set of attributes that are required for every user in the UserPool to signup.
```

Author: fixed

```ts
import { App, Stack } from 'aws-cdk-lib';
import { AmplifyAuth } from '@aws-amplify/auth-construct';

const app = new App();
const stack = new Stack(app, 'Auth', {
  loginWith: { email: true },
  userAttributes: {
    address: {
      mutable: false,
    },
    familyName: {
      required: true,
    },
  },
});

> new AmplifyAuth(stack, 'Auth');

Reviewer: Looks like there is a typo in the code. Props are sent to the stack 🙃

Author: ah copy paste error

# AmplifyAuth Construct

This package vends an L3 CDK Construct wrapping the behavior of Amplify Auth. This enables quick development and interation with AWS Cognito.

The primary way to use this construct is to invoke it with a configuration object. You can declare the individual settings for your authentication by passing them as properties to the AmplifyAuth construct.

> Note: only a single instance of the `AmplifyAuth` construct can be invoked within a CDK synthesis at this point in time.

Reviewer 3: Do we validate this at runtime ?

Author: yes it will fail at cdk synth 
![Uploading image.png…]()

## Examples

### Simple email login with default settings

In this example, we create a simple stack with email login enabled (by default). Deploying this will create a UserPool, UserPoolClient, IdentityPool, and Authenticated/Unauthenticated Roles.

We create a cdk App and Stack for this example. You may be deploying this to a custom stack, but this is purely illustrative for a concise demo.

```ts
import { App, Stack } from 'aws-cdk-lib';
import { AmplifyAuth } from '@aws-amplify/auth-construct';

const app = new App();
const stack = new Stack(app, 'Auth');

new AmplifyAuth(stack, 'Auth');

> ```

Reviewer 3: could you create an issue in GH to find a solution to this problem: "How do we keep these samples in sync?".
One of the options is to somehow keep them in TS files to keep them under compiler eye and have some automation that detect diff between MD file and snippets.

Author: I have the same question regarding all of docs

Reviewer 3: docs are going to be harder problem to solve outside of the repo, but if we build mechanism here we could then spread it to docs.
